### PR TITLE
bpo-44002: Switch to lru_cache in urllib.parse.

### PR DESCRIPTION
**Pull request by gpshead**
urllib.parse now uses functool.lru_cache for its internal URL splitting and
quoting caches instead of rolling its own like its the 90s.

The undocumented internal Quoted class is now deprecated, for removal in 3.14.
No existing users were found.

clear_cache() is kept as twisted, gevent, and our own regression test runner were
found to be using it in their test suites.

https://bugs.python.org/issue44002

Automerge-Triggered-By: GH:gpshead

### CODE REVIEW

**Comment by gpshead** [comment]
I won't merge this until after the 3.10 branch is cut and all of the https://bugs.python.org/issue43882 fixes are in.

File: Lib/urllib/parse.py
Comment (rhettinger) [inline Lib/urllib/parse.py]: If it were me, I wouldn't specify a maxsize and just accept the default.   Also, there is *slight* advantage to making the maxsize a power of two.

File: Lib/urllib/parse.py
Comment (rhettinger) [inline Lib/urllib/parse.py:886]: Again, I would go with the default size.  The LRU cache isn't presized; rather, it grows as needed.  So, there is no benefit in selecting a smaller size because of the expected number of cases.  The purpose of the *maxsize* to prevent excessive memory use in the extreme cases.  So we choose the size mainly based on our tolerance worst case.  Since the size of the inputs and outputs for a quoter are very small, we could tolerate thousands of entries and never notice the memory consumption.

File: Lib/test/test_urlparse.py
Comment (rhettinger) [inline Lib/test/test_urlparse.py]: I would go directly to a DeprecationWarning.  We haven't had much luck with people ever seeing and benefitting from a PendingDeprecationWarning.

Comment (gpshead) [reply Lib/test/test_urlparse.py]: I'll defer deciding this until I do some OSS searching to see if I can even find _any_ users. (i bet not, these are obscure)

**Review by gpshead — COMMENTED** [review]
The test that's failing due to an unexpected BytesWarning in CI is annoying... that means we need to use `lru_cache(typed=True)` on the urlsplit decorator for safety to avoid other people who enable BytesWarning from seeing that warning.

File: Lib/urllib/parse.py
Comment (serhiy-storchaka) [inline Lib/urllib/parse.py:902]: This is a de-optimization. List comprehension is faster than generator expression.

Comment (gpshead) [reply Lib/urllib/parse.py:902]: sad and understandable, but it appears true.  i restored them.

**Comment by rhettinger** [comment]
It looks like the Travis CI failed only because of excess whitespace.  Running 'make patchcheck' would likely clear it.

File: Lib/test/test_urlparse.py
Comment (tirkarthi) [reply Lib/test/test_urlparse.py]: I can see two usages in grepping top 1000 PyPI packages.

```
$ rg -t py 'urllib.parse.*clear_cache'
twisted/src/twisted/web/test/test_http.py
17:from urllib.parse import clear_cache  # type: ignore[attr-defined]

gevent/src/gevent/testing/leakcheck.py
130:            sys.modules['urllib.parse'].clear_cache()
```

**Comment by orsenthil** [comment]
This final change, which addressed the review comments from the reviewers looks good to me.

File: Lib/test/test_urlparse.py
Comment (gpshead) [reply Lib/test/test_urlparse.py]: Thanks!  Based on that (and our own libregrtest use), i'll just keep clear_cache and not deprecate it.

**Comment by gpshead** [comment]
Status: awaiting a copy of https://github.com/python/cpython/pull/25853 to be merged into `main` before merging this one.

File: Lib/urllib/parse.py
Comment (akulakov) [inline Lib/urllib/parse.py:886]: I think comment is no longer needed? [Edit: on a second thought, might be good to have, please disregard.]

**Review by serhiy-storchaka — APPROVED** [review]

**Comment by orsenthil** [comment]
Excellent change. @gpshead :)

**Pull request merged**
